File: docs/relay_server.py
import asyncio
import websockets
from collections import defaultdict

# Store sessions: session_code -> set of websocket connections
sessions = defaultdict(set)

async def handle_client(websocket, path):
    session_code = path.strip('/')
    if not session_code:
        await websocket.close()
        return
    
    print(f"New connection to session: {session_code}")
    sessions[session_code].add(websocket)
    
    try:
        async for message in websocket:
            # Relay message to all other clients in the same session
            for client in sessions[session_code]:
                if client != websocket and not client.closed:
                    await client.send(message)
    finally:
        sessions[session_code].discard(websocket)
        if not sessions[session_code]:
            del sessions[session_code]
            print(f"Session {session_code} closed (no clients)")

async def main():
    port = 8080
    async with websockets.serve(handle_client, "0.0.0.0", port):
        print(f"WebSocket relay server running on port {port}")
        await asyncio.Future()  # run forever

if __name__ == "__main__":
    asyncio.run(main())

File: docs/relay-server.js
const WebSocket = require('ws');
const http = require('http');

const server = http.createServer();
const wss = new WebSocket.Server({ server });

// Store sessions: sessionCode -> Set of WebSocket connections
const sessions = new Map();

wss.on('connection', (ws, req) => {
    const sessionCode = req.url.slice(1); // Remove leading '/'
    console.log(`New connection to session: ${sessionCode}`);
    
    if (!sessionCode) {
        ws.close();
        return;
    }
    
    // Add client to session
    if (!sessions.has(sessionCode)) {
        sessions.set(sessionCode, new Set());
    }
    sessions.get(sessionCode).add(ws);
    
    ws.on('message', (message) => {
        // Relay message to all other clients in the same session
        const sessionClients = sessions.get(sessionCode);
        if (sessionClients) {
            sessionClients.forEach((client) => {
                if (client !== ws && client.readyState === WebSocket.OPEN) {
                    client.send(message);
                }
            });
        }
    });
    
    ws.on('close', () => {
        // Remove client from session
        const sessionClients = sessions.get(sessionCode);
        if (sessionClients) {
            sessionClients.delete(ws);
            if (sessionClients.size === 0) {
                sessions.delete(sessionCode);
                console.log(`Session ${sessionCode} closed (no clients)`);
            }
        }
    });
});

const PORT = process.env.PORT || 8080;
server.listen(PORT, () => {
    console.log(`WebSocket relay server running on port ${PORT}`);
});
